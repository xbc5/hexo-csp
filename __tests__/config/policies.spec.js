"use strict";
const Config = require("../../lib/config");

const policies = () => [
  {
    pattern: "^foo$",
    prod: {
      mode: "merge",
      directives: {
        ["default-src"]: ["https://prod-foo-default.com"],
      },
    },
    dev: {
      mode: "merge",
      directives: {
        ["default-src"]: ["https://dev-foo-default.com"],
      },
    },
  },
  {
    pattern: "^bar$",
    prod: {
      mode: "merge",
      directives: {
        ["default-src"]: ["https://prod-bar-default.com"],
      },
    },
    dev: {
      mode: "merge",
      directives: {
        ["default-src"]: ["https://dev-bar-default.com"],
      },
    },
  },
];

const factory = (p = policies()) =>
  new Config({
    csp: { policies: p },
  });

it("should return the specified policies as-is", async () => {
  const result = factory().policies;
  expect(result).toMatchSnapshot();
});

it("should return an empty array if there's no policies", async () => {
  const result = factory(null).policies;
  expect(result).toStrictEqual([]);
});

it("should return a clone", async () => {
  // WARN: the constructor clones the entire config, and "policies" getter also clones
  // so even if you remove the getter clone, this will still pass. Leave this in place
  // just in case.
  const p = policies();
  const result = factory(p).policies;

  p[0].pattern = "changed";
  p[0].prod.directives["default-src"].push("changed");
  p[0].prod.mode = "changed";

  expect(result).toStrictEqual(policies());
});
