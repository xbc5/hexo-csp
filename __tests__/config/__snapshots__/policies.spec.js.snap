// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return the specified policies as-is 1`] = `
Array [
  Object {
    "dev": Object {
      "directives": Object {
        "default-src": Array [
          "https://dev-foo-default.com",
        ],
      },
      "mode": "merge",
    },
    "pattern": "^foo$",
    "prod": Object {
      "directives": Object {
        "default-src": Array [
          "https://prod-foo-default.com",
        ],
      },
      "mode": "merge",
    },
  },
  Object {
    "dev": Object {
      "directives": Object {
        "default-src": Array [
          "https://dev-bar-default.com",
        ],
      },
      "mode": "merge",
    },
    "pattern": "^bar$",
    "prod": Object {
      "directives": Object {
        "default-src": Array [
          "https://prod-bar-default.com",
        ],
      },
      "mode": "merge",
    },
  },
]
`;
